aten/src/ATen/core/dispatch/Dispatcher.h

```diff
@@ -399,48 +424,139 @@ inline Return Dispatcher::callWithDispatchKey(const TypedOperatorHandle<Return(A
+        int64_t seq_num = -1;
+        // Setting sequence number in the Autograd case to associate
+        // the forward range with the coresponding Autograd's node
+        if (isIncludedInAlias(dispatchKey, DispatchKey::Autograd) && at::GradMode::is_enabled()) {
+          seq_num = at::sequence_number::peek();
+        }
+        if (guard.needsInputs()) {
+          torch::jit::Stack stack = impl::boxArgs(args...);
+          guard.before(op, stack, seq_num);
+        } else {
+          guard.before(op, seq_num);
+        }
+      }
+    }
+    // keeping the guard alive while executing the kernel
+    return kernel.template call<Return, Args...>(op, dispatchKeySet, std::forward<Args>(args)...);
   }
 #endif  // PYTORCH_DISABLE_PER_OP_PROFILING
-  return kernel.template call<Return, Args...>(op, std::forward<Args>(args)...);
+  return kernel.template call<Return, Args...>(op, dispatchKeySet, std::forward<Args>(args)...);
 }
 
 template<class Return, class... Args>
-inline Return Dispatcher::call(const TypedOperatorHandle<Return(Args...)>& op, Args... args) const {
+// Note: benchmarks showed that this function wasn't getting inlined during calls to at::empty
+C10_ALWAYS_INLINE Return Dispatcher::call(const TypedOperatorHandle<Return(Args...)>& op, Args... args) const {
   detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
-  auto dispatchKey = op.operatorIterator_->op.dispatchKeyExtractor()
-    .template getDispatchKeyUnboxed<Args...>(
+  auto dispatchKeySet = op.operatorIterator_->op.dispatchKeyExtractor()
+    .template getDispatchKeySetUnboxed<Args...>(
       DispatchKeySet::FULL,
       args...
     );
-  return callWithDispatchKey<Return, Args...>(op, dispatchKey, args...);
+  return callWithDispatchKeySet<Return, Args...>(op, dispatchKeySet, args...);
 }
 
 template<class Return, class... Args>
 inline Return Dispatcher::redispatch(const TypedOperatorHandle<Return (Args...)>& op, DispatchKey currentDispatchKey, Args... args) const {
   detail::unused_arg_(args...);  // workaround for a false-positive warning about unused parameters in gcc 5
-  auto dispatchKey = op.operatorIterator_->op.dispatchKeyExtractor()
-    .template getDispatchKeyUnboxed<Args...>(
+  auto currentDispatchKeySet = op.operatorIterator_->op.dispatchKeyExtractor()
+    .template getDispatchKeySetUnboxed<Args...>(
       DispatchKeySet(DispatchKeySet::FULL_AFTER, currentDispatchKey),
       args...
     );
   // do not use RecordFunction on redispatch
-  const KernelFunction& kernel = op.operatorIterator_->op.lookup(dispatchKey);
-  return kernel.template call<Return, Args...>(op, std::forward<Args>(args)...);
+  const KernelFunction& kernel = op.operatorIterator_->op.lookup(currentDispatchKeySet.highestPriorityTypeId());
+  return kernel.template call<Return, Args...>(op, currentDispatchKeySet, std::forward<Args>(args)...);
+}
+
+// TODO: temp
```

Comment:
I see temp: what's the plan here? I'm not sure if I'd want a function like this at all, but assuming we do, I wouldn't call it "redispatch" because it is quite different from dispatch key only redispatch (that one will handle the masking for you so that you end up dispatching to the NEXT dispatch key; that's why this is specified as CURRENT dispatch key. But in your calls to the new redispatch method, you mask at call site; so the API contract here is not the same.

Also, even if all the above weren't true, it would still be better to keep these different names, since there isn't really any good reason to overload an internal API.

Reply:
Completely right, I should have gone through my `// TODO: temp`'s more carefully before the next round 😛 


> I'm not sure if I'd want a function like this at all

Just making sure I understand your motivation behind this. Isn't the "essence" of this function pretty central to the PR? i.e. a new API that takes in a fully pre-computed DispatchKeySet.

The current name is definitely not good. Naming things is hard- I'm opting for `callWithPrecomputedDispatchKeySet()`. Let me know if you think of something better.

Reply:
> Just making sure I understand your motivation behind this. Isn't the "essence" of this function pretty central to the PR? i.e. a new API that takes in a fully pre-computed DispatchKeySet.

I'm not sure about this. But it seems to me that there are two ways you could go about doing this: you could ask call sites to handle masking out the unnecessary key, or you could incorporate it as part of this function. In the latter case, the function *would* be like redispatch, and you'd pass it both the pre-computed dispatch key set, as well as the "current" key which it would use to mask out. Or you could do a clz test to to get around having to provide the current key at all.

Upon further reflection, I think this is all up to your personal taste. I'm just pointing out that there's more than one way to do it. One thing to think about is how to minimize the probability people will misuse the API.